drivers/SPI.cpp

```diff
@@ -75,69 +141,78 @@ void SPI::frequency(int hz)
     // If changing format while you are the owner then just
     // update frequency, but if owner is changed then even frequency should be
     // updated which is done by acquire.
-    if (_owner == this) {
-        spi_frequency(&_spi, _hz);
+    if (_peripheral->owner == this) {
+        spi_frequency(&_peripheral->spi, _hz);
     } else {
         _acquire();
     }
     unlock();
 }
 
-SPI *SPI::_owner = NULL;
-SingletonPtr<PlatformMutex> SPI::_mutex;
-
-// ignore the fact there are multiple physical spis, and always update if it wasn't us last
-void SPI::aquire()
-{
-    lock();
-    if (_owner != this) {
-        spi_format(&_spi, _bits, _mode, 0);
-        spi_frequency(&_spi, _hz);
-        _owner = this;
-    }
-    unlock();
-}
-
 // Note: Private function with no locking
 void SPI::_acquire()
 {
-    if (_owner != this) {
-        spi_format(&_spi, _bits, _mode, 0);
-        spi_frequency(&_spi, _hz);
-        _owner = this;
+    if (_peripheral->owner != this) {
+        spi_init(&_peripheral->spi, _mosi, _miso, _sclk, _hw_ssel);
+        spi_format(&_peripheral->spi, _bits, _mode, 0);
+        spi_frequency(&_peripheral->spi, _hz);
+        _peripheral->owner = this;
     }
 }
 
 int SPI::write(int value)
 {
-    lock();
-    _acquire();
-    int ret = spi_master_write(&_spi, value);
-    unlock();
+    select();
+    int ret = spi_master_write(&_peripheral->spi, value);
+    deselect();
     return ret;
 }
 
 int SPI::write(const char *tx_buffer, int tx_length, char *rx_buffer, int rx_length)
 {
-    lock();
-    _acquire();
-    int ret = spi_master_block_write(&_spi, tx_buffer, tx_length, rx_buffer, rx_length, _write_fill);
-    unlock();
+    select();
+    int ret = spi_master_block_write(&_peripheral->spi, tx_buffer, tx_length, rx_buffer, rx_length, _write_fill);
+    deselect();
     return ret;
 }
 
+void SPI::_set_ssel(int val)
+{
+    if (_sw_ssel.is_connected()) {
+        _sw_ssel = val;
+    }
+}
+
 void SPI::lock()
 {
-    _mutex->lock();
+    _peripheral->mutex->lock();
+}
+
+void SPI::select()
+{
+    lock();
+    if (_select_count++ == 0) {
+        _acquire();
+        _set_ssel(0);
+    }
 }
 
 void SPI::unlock()
 {
-    _mutex->unlock();
+    _peripheral->mutex->unlock();
+}
+
+void SPI::deselect()
+{
+    if (--_select_count == 0) {
+        _set_ssel(1);
+    }
+    unlock();
 }
 
 void SPI::set_default_write_value(char data)
 {
+    // this does not actually need to lock the peripheral.
```

Comment:
So could we remove lock/unlock?

Reply:
Not necessarily - the lock is also functioning as the "API lock". If the SPI object is itself supposed to be thread-safe (as opposed to just thread safety for multiple SPI objects accessing one bus), then there needs to be synchronisation on its members.  It currently seems that such objects are trying to be locally thread-safe, but maybe that needs more thought.